src/Illuminate/Database/Eloquent/Builder.php

```diff
@@ -1139,10 +1139,15 @@ protected function addUpdatedAtColumn(array $values)
 
         $column = $this->model->getUpdatedAtColumn();
 
-        $values = array_merge(
-            [$column => $this->model->freshTimestampString()],
-            $values
-        );
+        if (! array_key_exists($column, $values)) {
+            $timestamp = $this->model->freshTimestampString();
+
+            if ($this->model->hasCast($column)) {
+                $timestamp = $this->model->newInstance([$column => $timestamp])->getAttributes()[$column];
+            }
+
+            $values = array_merge([$column => $timestamp], $values);
+        }
```

Comment:
The outer `array_key_exists` check is a micro optimisation here, as per the performance section of the description. It can be safely removed if we don't want it.